mono/metadata/icall.c

```diff
@@ -331,21 +334,17 @@ array_set_value_impl (MonoArrayHandle arr, MonoObjectHandle value, guint32 pos,
 		break;
 	}
 
-	MonoObjectHandle inst;
-	inst = mono_object_handle_isinst (value, ec, error);
-	goto_if_nok (error, leave);
-	gboolean castOk;
-	castOk = !MONO_HANDLE_IS_NULL (inst);
+	castOk = mono_class_is_assignable_from_internal (ec, vc);
 
 	if (!m_class_is_valuetype (ec)) {
 		if (!castOk)
 			INVALID_CAST;
-		MONO_HANDLE_ARRAY_SETREF (arr, pos, value);
+		MONO_HANDLE_ARRAY_SETREF (arr_handle, pos, value_handle);
 		goto leave;
 	}
 
 	if (castOk) {
-		va = (gpointer*)mono_object_handle_pin_unbox (value, &value_gchandle);
+		va = (gpointer*) mono_object_unbox_internal (value);
```

Comment:
the contents of this `if (castOk) { ... }` block would make another good no safepoints region (together with an assignment to `ea` from `arr_handle`.